src/internal/operators/publishReplay.ts

```diff
@@ -9,6 +9,85 @@ export function publishReplay<T>(bufferSize?: number, windowTime?: number, sched
 export function publishReplay<T, O extends ObservableInput<any>>(bufferSize?: number, windowTime?: number, selector?: (shared: Observable<T>) => O, scheduler?: SchedulerLike): OperatorFunction<T, ObservedValueOf<O>>;
 /* tslint:enable:max-line-length */
 
+/**
+ * Returns a ConnectableObservable, which is a variety of Observable that waits until its connect method is called
+ * before it begins emitting items to those Observers that have subscribed to it. Internally a ReplaySubject is created
+ * to support caching items for replay to Observers that subscribe after emission begins. Cached values can also be delayed
+ * by a given amount of time.
+ *
+ * <span class="informal">Makes a cold Observable hot and allows for caching of source values</span>
+ *
+ * ![](publishReplay.png)
+ *
+ * ## Examples
+ * Make source$ hot by applying publishReplay operator, and cache 1 value in a ReplaySubject
+ * for any observers that subscribe after the source begins emitting.
+ * ```ts
+ * import { of, zip, interval, merge } from "rxjs";
+ * import { map, publishReplay, tap, refCount } from "rxjs/operators";
+ *
+ * const source$ = zip(interval(1000), of(1, 2, 3, 4, 5, 6, 7, 8, 9)).pipe(
+ *  map(values => values[1]),
+ *    publishReplay(1),
+ *    refCount()
+ *  );
+ *
+ * source$.subscribe(a => console.log(`Stream 1 ${a}`));
+ * setTimeout(() => source$.subscribe(a => console.log(`Stream 2 ${a}`)), 2500);
+ *
+ * // Stream 1 Results every one second
+ * // Stream 2 Begins after 2.5 seconds, then matches Stream 1
+ * // Stream 1: 1
+ * // Stream 1: 2
+ * // Stream 2: 2 After 0.5 seconds, the latest value from source is emitted
+ * // Stream 1: 3
+ * // Stream 2: 3 Now the streams match
+ * // ...
+ * // Stream 1: 9
+ * // Stream 2: 9
+ *
+ *
+ * // publishReplay can take an optional second argument windowTime to delay the emission of the cached values
+ * // to observers that subscribe after the source begins.
+ * import { of, zip, interval, merge } from "rxjs";
+ * import { map, publishReplay, tap, refCount } from "rxjs/operators";
+ *
+ * const source$ = zip(interval(1000), of(1, 2, 3, 4, 5, 6, 7, 8, 9)).pipe(
+ *  map(values => values[1]),
+ *    publishReplay(1, 500),
+ *    refCount()
+ *  );
+ *
+ * source$.subscribe(a => console.log(`Stream 1 ${a}`));
+ * setTimeout(() => source$.subscribe(a => console.log(`Stream 2 ${a}`)), 2600);
+ *
+ * // Stream 1 Results every one second
+ * // Stream 2 Begins after 3.1 seconds, then matches Stream 1
+ * // Stream 1: 1
+ * // Stream 1: 2
+ * // Stream 1: 3
+ * // Stream 2: 3 after 0.1 seconds, the latest value from source is emitted
+ * // Stream 1: 4
+ * // Stream 2: 4
+ * // ...
+ * // Stream 1: 9
+ * // Stream 2: 9
+ * ```
+ *
```

Comment:
I love the work here so far. Ironically, though, you haven't touched on the use case I'd most like to see people use that doesn't involve `refCount` or `ConnectableObservable` at all. You can multicast with a selection function:

```ts
of(1, 2, 3, 4).pipe(
  publishReplay(3, 3000, shared$ => merge(
    shared$.pipe(delay(1000), map(v => `one: ${v}`)),
    shared$.pipe(delay(300), map(v => `two ${v}`)),
  ))
  .subscribe({
    next: x => console.log(x),
    complete: () => console.log('done'),
  })

// Output
// one 1
// one 2
// one 3
// two 1
// one 4
// two 2
// two 3
// two 4
// done
```

Reply:
A good point! Admittedly I have not used it that way before and I am glad to have learned it. Thanks a bunch for all the suggestions and the example, and good eye on the code style issues. I will work on these changes tonight and get them up for your review when you have the time! @benlesh

Reply:
Okay I know it's been awhile this, I apologize! I have been in the middle of some life changes. But finally got back around to this. I think I am missing a piece of this somewhere...

So I took the example you had above and played around with it a bit to try to get a better understanding and I came up with this example:

`
import { merge, interval } from 'rxjs';
import { publishReplay, map, delay, takeWhile, publish } from 'rxjs/operators';

const source$ = interval(1000).pipe(
  takeWhile(x => x < 5),
  publishReplay(
    1,
    0,
    shared$ => merge(
      shared$.pipe(delay(1000), map(v => `Stream 1: ${v}`)),
      shared$.pipe(delay(500), map(v => `Stream 2: ${v}`)),
    ),
  ),
 );

 source$.subscribe({
  next: x => console.log(`Sub 1: ${x}`),
  complete: () => console.log('Sub 1 done'),
 });

setTimeout(() => {
   source$.subscribe({
    next: x => console.log(`Sub 2: ${x}`),
    complete: () => console.log('Sub 2 done'),
  });
}, 3000);

// Sub 1: Stream 2: 0
// Sub 1: Stream 1: 0
// Sub 1: Stream 2: 1
// Sub 1: Stream 1: 1
// Sub 1: Stream 2: 2
// Sub 1: Stream 1: 2
// Sub 1: Stream 2: 3
// Sub 1: Stream 2: 0
// Sub 2: Stream 2: 3
// Sub 2: Stream 2: 0
// Sub 1: Stream 1: 3
// Sub 2: Stream 1: 3
// Sub 2: Stream 1: 0
// Sub 1: Stream 1: 0
// Sub 1: Stream 2: 4
// Sub 1: Stream 2: 1
// Sub 2: Stream 2: 4
// Sub 2: Stream 2: 1
// Sub 1: Stream 1: 4
// Sub 1: Stream 1: 1
// Sub 2: Stream 1: 4
// Sub 2: Stream 1: 1
// Sub 1 done
// Sub 2 done
`

This is different from what I expected to see, so I was hoping for some clarification!
The second subscription picks up the replayed emissions from the selector as I would expect (`Sub 2: Stream 2: 3` and `Sub 2: Stream 1: 3` emit when the second subscription is created) but that subscription also seems to restart the whole stream over again and BOTH subscriptions restart back at 0 again after the second subscription begins (and they both finish out the original stream to 4). What part of this causes the streams to restart like that?